src/mon/KVMonitor.cc

```diff
@@ -0,0 +1,419 @@
+}
+
+const string KV_PREFIX = "mon_config_key";
+
+const int MAX_HISTORY = 50;
+
+
+static bool is_binary_string(const string& s)
+{
+  for (auto c : s) {
+    // \n and \t are escaped in JSON; other control characters are not.
+    if ((c < 0x20 && c != '\n' && c != '\t') || c >= 0x7f) {
+      return true;
+    }
+  }
+  return false;
+}
+
+
+KVMonitor::KVMonitor(Monitor &m, Paxos &p, const string& service_name)
+  : PaxosService(m, p, service_name) {
+}
+
+void KVMonitor::init()
+{
+  dout(10) << __func__ << dendl;
+}
+
+void KVMonitor::create_initial()
+{
+  dout(10) << __func__ << dendl;
+  version = 0;
+  pending.clear();
+}
+
+void KVMonitor::update_from_paxos(bool *need_bootstrap)
+{
+  if (version == get_last_committed()) {
+    return;
+  }
+  version = get_last_committed();
+  dout(10) << __func__ << " " << version << dendl;
+  //check_all_subs();
+}
+
+void KVMonitor::create_pending()
+{
+  dout(10) << " " << version << dendl;
+  pending.clear();
+}
+
+void KVMonitor::encode_pending(MonitorDBStore::TransactionRef t)
+{
+  dout(10) << " " << (version+1) << dendl;
+  put_last_committed(t, version+1);
+
+  // record the delta for this commit point
+  bufferlist bl;
+  encode(pending, bl);
+  put_version(t, version+1, bl);
+  
+  // make actual changes
+  for (auto& p : pending) {
+    string key = p.first;
+    if (p.second) {
+      dout(20) << __func__ << " set " << key << dendl;
+      t->put(KV_PREFIX, key, *p.second);
+    } else {
+      dout(20) << __func__ << " rm " << key << dendl;
+      t->erase(KV_PREFIX, key);
+    }
+  }
+}
+
+version_t KVMonitor::get_trim_to() const
+{
+  // we don't need that many old states, but keep a few
+  if (version > MAX_HISTORY) {
+    return version - MAX_HISTORY;
+  }
+  return 0;
+}
+
+void KVMonitor::get_store_prefixes(set<string>& s) const
+{
+  s.insert(service_name);
+  s.insert(KV_PREFIX);
+}
+
+void KVMonitor::tick()
+{
+  if (!is_active() || !mon.is_leader()) {
+    return;
+  }
+  dout(10) << __func__ << dendl;
+}
+
+void KVMonitor::on_active()
+{
+}
+
+
+bool KVMonitor::preprocess_query(MonOpRequestRef op)
+{
+  switch (op->get_req()->get_type()) {
+  case MSG_MON_COMMAND:
+    try {
+      return preprocess_command(op);
+    } catch (const bad_cmd_get& e) {
+      bufferlist bl;
+      mon.reply_command(op, -EINVAL, e.what(), bl, get_last_committed());
+      return true;
+    }
+  }
+  return false;
+}
+
+bool KVMonitor::preprocess_command(MonOpRequestRef op)
+{
+  auto m = op->get_req<MMonCommand>();
+  std::stringstream ss;
+  int err = 0;
+
+  cmdmap_t cmdmap;
+  if (!cmdmap_from_json(m->cmd, &cmdmap, ss)) {
+    string rs = ss.str();
+    mon.reply_command(op, -EINVAL, rs, get_last_committed());
+    return true;
+  }
+  string format;
+  cmd_getval(cmdmap, "format", format, string("plain"));
+  boost::scoped_ptr<Formatter> f(Formatter::create(format));
+
+  string prefix;
+  cmd_getval(cmdmap, "prefix", prefix);
+  string key;
+  cmd_getval(cmdmap, "key", key);
+
```

Comment:
are permissions already checked elsewhere? I'd expect those checks here

Reply:
also in prepare_command()

Reply:
they're checked in the caller at https://github.com/ceph/ceph/blob/master/src/mon/Monitor.cc#L3422